### ISSUE TITLE
[BUG] canonical URL with multiple domains but same siteroot

### ISSUE THREAD

**Issue by AndreasA** [mblaschke/TYPO3-metaseo#110]
If you have multiple domain entries for the same siteroot, the canonical tag should always use only one of those domains which would be the "first" domain in the list.
Of course, in most cases one creates a redirect anyway but...

e.g. if you have something like:
www.example.org
foo.example.org
and both have the same site root, let's say page uid 1
Then the canonical tags should all point to www.example.org.
Currently at foo.example.org the would still point to foo.example.org - except maybe if one uses a fix baseURL or similiar.

**Comment by thomaszbz** [comment]
I'm not sure if this is the case in every scenario. What if you have one subdomain per language? Then you'd also want to have one canonical url per translation. In this scenario you would also have the same site root for multiple domains.

Plus, simply choosing the "first" domain in a list seems to be a limitation. However, you could come up with a setting which forces a domain by configuration, still allowing reasonable defaults.

The other way round I prefer to avoid duplicate content which basically means that you decide which unique url (inlcuding the protocol) is used for unique content. Any other url would be 301ed to the unique content. I know that this is not possible for every scenario, e.g. if you want to provide content via multiple protocols, e.g. http and https.

That said, what would be a concept you think is reasonable?

**Comment by AndreasA** [comment]
Hmm. true all the time might be wrong.
Maybe an optional feature where one specifies which domain to use per root page?

As mentioned in most cases one would use a redirect anyway so it will not happen that often but I still think the extension should handle this.

**Comment by AndreasA** [comment]
However, I think the first in the list should be enough as the domain entries are sortable.

Another possibility altogether:
We could add a checkbox to the domain entries like "allowed" domains for this root page.
If the domain is not in the allowed domains it will use one of the allowed domains, e.g. first or it could have another checkbox like "preferred" domain.

**Comment by thomaszbz** [comment]
Yes, it's just not clear what would be "right". This can only be implemented when a setting says how it should be. In any case we should have a good concept for this issue before we (or you) start coding on this issue.

Sorting the domains is not a good approach, e.g. if some wants an explicit domain. I'd expect this would also be the normal case.

I'm wondering what you want to do with a set of "allowed" domains. If I got you right, you want to use explicitly _one_ protocol and domain for canonical url generation. Basically this means that there should not be a set of "allowed" domains.

You could also argue that duplicate content could be the output of multiple TYPO3 pages. This should also be taken in consideration.

In the end you could still use the flexible approach of MetaSEO which allows to set a url directly in the SEO section of page properties. This would also allow external urls, which, however, is not nice for a multidomain instance.

But indeed, we could talk about forcing a protocol and domain for canonical url generation, using inheritance of configuration. This could also apply for internal links by overwriting domain information.

**Comment by AndreasA** [comment]
Of cousre, how to implement it is up to a concept first.
There are actually multiple possibilities.
Not sure what the best way is though, e.g. using the sys_domain entries is limiting, if e.g. for a certain sub-page one wants to use a different preferntial domain.
Maybe TypoScript configuration would be the best way to go...

**Comment by AndreasA** [comment]
Allowed was just an idea, doesn't mean it is the best way but e.g. I have www.example.org and en.example.org in allowed domains, now if I get xzy.example.org that then produces a canonical tag to www.example.org because it is not allowed and www.example.org is preferred.

But as mentioned it is limiting in some ways and I acutally think the allowed domain concept is to complex. I think setting the wanted domain might be better. Not sure, how to configure it the best way though.
So I am putting this up to discussion anyway.

**Comment by thomaszbz** [comment]
TypoScript alone would conflict with user settings in the SEO tabs. We should avoid avoid that a user sets something in the SEO tab and is wondering why it does not have any effect.

**Comment by thomaszbz** [comment]
@AndreasA Have a look at #141 for the protocol preference in respect to cononical urls. That's one part of the story.

**Comment by AndreasA** [comment]
Thx. I wanted to take a look at it myself but other things took precedence.

**Comment by thomaszbz** [comment]
@AndreasA I just stumbled upon a feature in 8.7 LTS: `Always prepend this domain in links`. Not sure if this is new in 8LTS.

With some luck, this can be used to set a fixed domain. With some luck, metaseo even uses that. I did not test it out, however. Would be great if someone can report back what it does.

![auswahl_183](https://cloud.githubusercontent.com/assets/11200727/24844335/29aa7e16-1daa-11e7-8d49-8805915cd388.png)
![auswahl_182](https://cloud.githubusercontent.com/assets/11200727/24844337/2dc72cec-1daa-11e7-9a5e-7b07363d150c.png)
